File: src/main/java/ru/mrgrd56/api/common/provider/BeanProvider.java
package ru.mrgrd56.api.common.provider;

import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class BeanProvider<T extends Providable> {
    private final Map<String, T> beansMap;

    protected BeanProvider(List<T> beans) {
        this.beansMap = beans.stream()
                .collect(Collectors.toMap(Providable::getName, Function.identity()));
    }

    public T get(String name) {
        return beansMap.get(name.toLowerCase());
    }
}

File: src/main/kotlin/ru/mrgrd56/api/common/provider/BeanProvider.kt
package ru.mrgrd56.api.common.provider

import ru.mrgrd56.api.exception.NotFoundException
import java.util.*

open class BeanProvider<T : Providable> protected constructor(beans: List<T>) {
    private val beansMap: Map<String, T>

    init {
        beansMap = beans.associateBy { it.name }
    }

    @Throws(NotFoundException::class)
    operator fun get(name: String): T {
        return beansMap[name.lowercase(Locale.getDefault())] ?: throw NotFoundException("Bean not found")
    }
}
